### ISSUE TITLE
_InputArray::~_InputArray() is not virtual

### ISSUE THREAD

**Issue by csukuangfj** [opencv/opencv#11978]
##### System information (version)

- OpenCV => latest

##### Detailed description

See the code
https://github.com/opencv/opencv/blob/fa66c6b797416292e2c6befb60b34b6ba0420499/modules/core/include/opencv2/core/mat.hpp#L241

https://github.com/opencv/opencv/blob/fa66c6b797416292e2c6befb60b34b6ba0420499/modules/core/include/opencv2/core/mat.hpp#L278

https://github.com/opencv/opencv/blob/fa66c6b797416292e2c6befb60b34b6ba0420499/modules/core/include/opencv2/core/mat.hpp#L363

**Comment by alalek** [comment]
`InputArray` / `OutputArray` / `InputOutputArray` object is supposed to be crated and destroyed by caller (in the same scope).
Casts and other places of destruction are not expected here.

**Comment by csukuangfj** [comment]
A good design should make the usage less error prone.

**Comment by sergiud** [comment]
You can't use for instance `std::vector` safely as base class either. Does this mean `std::vector` design is flawed? Many if not most would disagree. Consequently, you cannot shift the burden of providing a "good design" completely to the library. The user is ought to apply common sense to a certain degree as well.

**Comment by csukuangfj** [comment]
@sergiud 
thank you for the comment.

How did you come about the following statement
> You can't use for instance std::vector safely as base class either.

It is not uncommon to see people inheriting from `std::vector`. 
For example, see here 
https://github.com/MarcToussaint/rai/blob/cfdf57d7151e7c862d375efc20ac002892bae073/rai/Core/array.h#L71

If you have designed a class such that users have to follow some certain rules while using it, otherwise memory leakage will occur. As far as I am concerned, I won't consider it as a good practice. Insterad, I think it is a bug. A good design should make it hard to misuse.

**Comment by sergiud** [comment]
Just because it's not uncommon that people inherit from `std::vector`, it does not mean that's good design. Conversely, if you inherit from `std::vector` you are expected to know what you are doing. Same rules apply to `cv::InputArray` which are simply container adapters are not supposed to be used in class hierarchy (i.e., being referenced through base class pointers). Thus these adaptors don't need virtual destructors.

**Comment by csukuangfj** [comment]
@sergiud
What are the problems if it were made virtual?

**Comment by sergiud** [comment]
ABI change, enforcement of RTTI + payload (vtable).

**Comment by csukuangfj** [comment]
@sergiud 

1. OpenCV 4.X is guaranteed to be incompatible with OpenCV3.X, so ABI changes are not issues.
2. Make the destructor virtual will incur only several bytes for creating the vtable. Since only the destructor is virtual there is no runtime cost for average users.

The advantage I can think of now is that it is able to prevent possible memory leak for careless users.

**Comment by sergiud** [comment]
You ignore enforcement of RTTI which is a deal breaker.

**Comment by csukuangfj** [comment]
Please refer to **OE 4. OpenCV 4** https://github.com/opencv/opencv/wiki/OE-4.-OpenCV-4

> Disclaiming binary compatibility and moving to source-level compatibility. Several years ago we stated that OpenCV will stay binary-compatible between major releases. Technically, it was a necessary thing for OpenCV Manager for Android, but it also was seemingly useful thing for binary OpenCV packages for Linux. Apparently, neither of these two is that important. App developers do not rely on independently-updated package and Linux distro maintainers prefer to do just some security updates. For them the difference between, say, 3.0 and 3.1 is too radical. So we can have more freedom and yet provide a reasonable level of compatibility by relaxing the "binary compatibility" claim to "source compatibility".

**Comment by sergiud** [comment]
I'm not sure what you are trying to say. Please address my points directly, thanks.

**Comment by csukuangfj** [comment]
@sergiud 
Since OpenCV4.x does not require ABI compatibility, what did you mean of **enforcement of RTTI**?

Is there any similarity between **InputArray** and **std::vector**? I have no idea which class inside STL inherits **std::vector**. But **InputArray** is inherited by **OutputArray**.

**Comment by sergiud** [comment]
The question is whether it is possible to safely disable RTTI (induced by the virtual destructors) without any problems. Adding virtual destructor unconditionally introduces an overhead in terms of a vtable which most of the users would not benefit from. So it is mostly a wasted effort to prevent a misuse which even does not occur that frequently.

I also believe it should not matter what class InputArray inherits from. This is an implementation detail users should not care about.

**Comment by alalek** [comment]
We would like to avoid adding extra overhead without solving a real problems (incorrect / "unexpected" usage is not a case).

**Comment by sergiud** [comment]
In addition, a clean solution for this "problem" would involve either making the destructor of adaptor base classes protected or entirely avoiding the use of class hierarchies. Patching the destructor using the proposed solution neither constitutes a good design choice nor it is practicable.

**Comment by kinchungwong** [comment]
Per Exceptional C++, there are times when it is necessary to make a certain C++ class "non-inheritable" (not eligible for subclassing). Unfortunately, older C++ standard doesn't provide a way to prevent subclassing. The newer C++11 standard has the ```final``` keyword. In addition to that, there are several tricks which will limit subclassing to within a library's internal code base, but those will increase code clutter.

**Comment by vpisarev** [comment]
Indeed, `InputArray_` and `OutputArray_` are supposed to be "final" classes, meaning that users should not derive from them. We inherit `OutputArray_` from `InputArray_` just to reuse some of its functionality, like `getMat()`. By using these `*Array` classes in API we get much more convenience at the cost of some overhead. If we make their methods virtual (and making just the destructor virtual does not make sense, right? Then we should make `getMat()` and other methods virtual too), we might get slightly more convenience for very advanced users who would like to make their custom array types, at the cost of even bigger overhead for all users. Currently, `InputArray_` destructor call is optimized out, since the compiler knows it's empty. `InputArray::getMat()` call will be easier to predict since it's not virtual etc.
